File: 백준/Platinum/1028. 다이아몬드 광산/다이아몬드 광산.py
import sys

N, M = 0, 0
m = []
count_m = []

class Node:
    def __init__(self, x, y):
        self.x = x
        self.y = y

def find_dia():
    ret = 0
    for i in range(N):
        for j in range(M):
            min_val = min(count_m[i][j][0], count_m[i][j][1])
            if min_val <= ret:
                continue
            while min_val > ret:
                X = i + (min_val - 1)
                leftY = j - (min_val - 1)
                rightY = j + (min_val - 1)
                min_val -= 1
                if not m_chk(X, leftY) or not m_chk(X, rightY):
                    continue
                if count_m[X][rightY][0] < min_val + 1 or count_m[X][leftY][1] < min_val + 1:
                    continue
                min2 = min(count_m[X][leftY][1], count_m[X][rightY][0])
                m = min(min_val + 1, min2)
                if ret < m:
                    ret = m
    return ret

def make_count_m():
    q = []
    for i in range(N):
        q.append(Node(i, 0))
    for j in range(1, M):
        q.append(Node(N - 1, j))
    make(q, True)
    q = []
    for i in range(N):
        q.append(Node(i, M - 1))
    for j in range(M - 1):
        q.append(Node(N - 1, j))
    make(q, False)

def make(q, flag):
    bef = 0
    while q:
        bef = 0
        n = q.pop(0)
        ni = n.x
        nj = n.y
        while True:
            if not m_chk(ni, nj):
                break
            if m[ni][nj]:
                bef += 1
            else:
                bef = 0
            count_m[ni][nj][0 if flag else 1] = bef
            ni -= 1
            nj = nj + 1 if flag else nj - 1

def m_chk(x, y):
    return 0 <= x < N and 0 <= y < M

def input_data():
    global N, M, m, count_m
    N, M = map(int, input().split())
    m = [[False] * M for _ in range(N)]
    count_m = [[[0, 0] for _ in range(M)] for _ in range(N)]

    for i in range(N):
        s = input()
        for j in range(M):
            m[i][j] = s[j] == '1'

def main():
    input_data()
    make_count_m()
    ans = find_dia()
    print(ans)

if __name__ == "__main__":
    main()

File: 백준/Platinum/1028. 다이아몬드 광산/다이아몬드 광산.cc
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

int N, M;
vector<vector<bool>> map;
vector<vector<vector<int>>> countMap;

struct Node {
    int x, y;

    Node(int x, int y) : x(x), y(y) {}
};

bool mapChk(int x, int y) {
    if (x < 0 || y < 0 || x >= N || y >= M)
        return false;
    return true;
}

void make(queue<Node>& q, bool flag) {
    int bef = 0;
    while (!q.empty()) {
        bef = 0;
        Node n = q.front();
        q.pop();
        int ni = n.x;
        int nj = n.y;
        while (true) {
            if (!mapChk(ni, nj))
                break;
            if (map[ni][nj])
                bef++;
            else
                bef = 0;
            countMap[ni][nj][flag ? 0 : 1] = bef;
            ni -= 1;
            nj = nj + (flag ? 1 : -1);
        }
    }
}

void makeCountMap() {
    queue<Node> q;
    for (int i = 0; i < N; i++)
        q.push(Node(i, 0));
    for (int j = 1; j < M; j++)
        q.push(Node(N - 1, j));
    make(q, true);
    q = queue<Node>();
    for (int i = 0; i < N; i++)
        q.push(Node(i, M - 1));
    for (int j = 0; j < M - 1; j++)
        q.push(Node(N - 1, j));
    make(q, false);
}

void input() {
    cin >> N >> M;
    map.resize(N, vector<bool>(M, false));
    countMap.resize(N, vector<vector<int>>(M, vector<int>(2, 0)));

    for (int i = 0; i < N; i++) {
        string s;
        cin >> s;
        for (int j = 0; j < M; j++)
            map[i][j] = (s[j] == '1');
    }
}

int findDia() {
    int ret = 0;
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            int minVal = min(countMap[i][j][0], countMap[i][j][1]);
            if (minVal <= ret)
                continue;
            while (minVal > ret) {
                int X = i + (minVal - 1);
                int leftY = j - (minVal - 1);
                int rightY = j + (minVal - 1);
                minVal--;
                if (!mapChk(X, leftY) || !mapChk(X, rightY))
                    continue;
                if (countMap[X][rightY][0] < minVal + 1 || countMap[X][leftY][1] < minVal + 1)
                    continue;
                int min2 = min(countMap[X][leftY][1], countMap[X][rightY][0]);
                int m = min(minVal + 1, min2);
                if (ret < m)
                    ret = m;
            }
        }
    }
    return ret;
}

int main() {
    input();
    makeCountMap();
    int ans = findDia();
    cout << ans << endl;

    return 0;
}